doc/specs/#885 - Terminal Settings Model.md

```diff
@@ -78,13 +78,126 @@ This separation leaves `AppKeyBindings` with the responsibility of detecting and
+```c++
+// layers the Profile settings onto another profile
+void Profile::ApplyTo(Profile profile) const
+{
+    if (_fontSize)
+    {
+        FontSize(_fontSize);
+    }
+
+    // repeat for all settings
+}
+```
+
+This functionality will be useful when layering the `profile.defaults` onto dynamic profiles.
+
+As a result, the tracking and functionality of cascading settings is moved into the object model instead of keeping it as a json-only concept.
+
+#### Updates to CascadiaSettings
+
+As `CascadiaSettings` loads the settings model, it will create children for each component of the settings model and layer the new values on top of it. Thus, `LayerJson` will look something like this:
+```c++
+void CascadiaSettings::LayerJson(const Json::Value& json)
+{
+    _globals = _globals.CreateInheritedProfile();
+    _globals->LayerJson(json);
+
+    // repeat the same for Profiles/ColorSchemes...
+}
+```
+For `defaults.json`, `_globals` will now hold all of the values set in `defaults.json`. If any settings were omitted from the `defaults.json`, `_globals` will fallback to its parent (a `GlobalAppSettings` consisting purely of system-defined values).
+
+For `settings.json`, `_globals` will only hold the values set in `settings.json`. If any settings were omitted from `settings.json`, `_globals` will fallback to its parent (the `GlobalAppSettings` built from `defaults.json).
+
+This process becomes a bit more complex for `Profile` because it can fallback in the following order:
+1. `settings.json` profile
+2. `settings.json` `profiles.defaults`
+3. (if a dynamic profile) the hardcoded value in the dynamic profile generator
+4. `defaults.json` profile
+
+`CascadiaSettings` must do the following...
+1. load `defaults.json`
+   - append newly created profiles to `_profiles` (unchanged)
+2. load dynamic profiles
+   - append newly created profiles to `_profiles` (unchanged)
+3. load `settings.json` `profiles.defaults`
+   - layer `profile.defaults` onto `_profiles` (unchanged)
+   - construct a `Profile` from `profiles.defaults`. Save as `Profile _profileDefaults`.
+4. load `settings.json` `profiles.list`
+   - if a matching profile exists, `CreateInheritedProfile` from the matching profile, and layer the json onto the child.
+   - otherwise, `CreateInheritedProfile` from `_profileDefaults`, and layer the json onto the clone.
+
+Additionally, `_profileDefaults` will be exposed by `Profile CascadiaSettings::ProfileDefaults()`. This will enable [#7414](https://github.com/microsoft/terminal/pull/7414)'s implementation to spawn incoming commandline app tabs with the "Default" profile (as opposed to the "default profile").
+
+
+### CreateChild vs Copy
+
+Settings objects will have `CreateChild()` and `Copy()`. `CreateChild()` is responsible for creating a new settings object that inherits undefined values from its parent. `Copy()` is responsible for recreating the contents of the settings object, including a reference to a copied parent.

```

Comment:
Maybe explicitly mention that `Copy` is a deep copy

Reply:
BE EXTREMELY CAREFUL DEEP COPYING!!!

If you naively copy `_parent`, you will end up with one deep-copied object tree whose members inherit from _a totally different object tree_.

We actually need a version of `Copy` that _takes a new parent_. Pretend that it's a "git rebase".

Actually, it _is_ a git rebase ...

Reply:
Added a whole discussion on this in this section. We'll be fine for now. But this'll be a real issue when we take on profile inheritance (idk if there's an issue tracking that?)